Course work 2024

N1X41 opened the pull request:
Первое тестирование

N1X41 committed:
feat: Запуск репозитория на гите

Первая версия с примитивной базовой логикой и почти полностью реализованный UI

nickfirst41 committed:
feat: Функция поиска идентичного узла

N1X41 committed:
feat: Улучшение вычислений и условий закрытия ветви

Добавил цикл перехода по списку, сократил количество переходов, добавил проверку по глубине

N1X41 committed:
feat: Сокращение и разбиение

nickfirst41 committed:
feat: Чекбокс отображения только решения

TonitaN commented on course_work_fkm/lib/classes/equation.dart:
```diff
@@ -0,0 +1,87 @@
+
+  /// Уравнение неразрешимо
+  bool isFalse() {
+    return ((left.length != right.length && left.length * right.length == 0) ||
```
Почему? Может быть и непустая часть равна пустой, если непустая состоит только из переменных.

N1X41 replied:
fix

TonitaN commented on course_work_fkm/lib/classes/equation.dart:
```diff
@@ -0,0 +1,87 @@
+  /// Уравнение неразрешимо
+  bool isFalse() {
+    return ((left.length != right.length && left.length * right.length == 0) ||
+        (left[0].toLowerCase() != left[0] &&
```
Лучше это сделать через обертки isVariable и isLetter

N1X41 replied:
fix

TonitaN commented on course_work_fkm/lib/classes/equation.dart:
```diff
@@ -0,0 +1,87 @@
+    // Убираем одинаковые символы с начала
+    while (left.length > 1 && right.length > 1 && 
+           left[0] == right[0] && 
+           left[0].toUpperCase() == left[0]) {
```
А одинаковые переменные не сокращаем?

N1X41 replied:
fix

TonitaN commented on course_work_fkm/lib/classes/equation.dart:
```diff
@@ -0,0 +1,87 @@
+        String leftSub = left.substring(0, i);
+        String rightSub = right.substring(0, i);
+        // Проверяем, равны ли подстроки по набору символов
+        if (Set.from(leftSub.split('')..sort()).toString() == Set.from(rightSub.split('')..sort()).toString()) {
```
Но разбиение может работать и на неравных подстроках. Например, если
`AxBy=xBAy`
Тут `Ax=xB` - подуравнение, но как отсортированные строки они не будут равны.
Вот если по переменным только - да, должны быть равны как отсортированные строки. + добавить равенство по длине обычное, тогда ваш метод будет работать.

N1X41 replied:
fix

TonitaN commented on course_work_fkm/lib/functions/for_graph.dart:
```diff
@@ -0,0 +1,223 @@
+
+    // Проверяем, что нет идентичного уравнения
+    for (Equation equat in equations) {
+      if (newNode.equations.every((eq) => eq != equat) &&
```
А коммутативность равенства не учитывается?

TonitaN replied:
Кстати, логично учесть его прямо в классе уравнения.

N1X41 replied:
fix

TonitaN reviewed (changes requested):
Правим базовую логику, а потом созваниваемся по поводу улучшения стратегии разбора.

N1X41 committed:
feat: Фикс по ревью

N1X41 committed:
feat: UI + логика

Добавлены иконки, показывающие причину закрытия ветки. Убраны неэффективные правила

N1X41 committed:
feat: Подстановки

TonitaN commented on course_work_fkm/lib/functions/for_nodes.dart:
```diff
@@ -1,7 +1,119 @@
+  // Подстановка с максимальным сокращением длины
+  Map<Equation, int> substitution = {};
+  for (int i = 0; i < equations.length; i++) {
+    if (equations[i].left != equations[i].right) {
```
Такие уравнения вообще сразу должны быть упрощены в ноль

TonitaN commented on course_work_fkm/lib/functions/for_nodes.dart:
```diff
@@ -1,7 +1,119 @@
+  Map<Equation, int> substitution = {};
+  for (int i = 0; i < equations.length; i++) {
+    if (equations[i].left != equations[i].right) {
+      if (equations[i].left.length > equations[i].right.length && _hasVars(equations[i].left)) {
```
Опять к вопросу о нормальных формах (распределение по длинам частей уравнения)

TonitaN commented on course_work_fkm/lib/functions/for_graph.dart:
```diff
@@ -214,6 +215,10 @@ void createNodeByRule(Graph graph, int index, Rule rule) {
+  newNode.equations = makeSubstitution(newNode.equations);
+
+  // Сортируем уравнения
   newNode.equations.sort(
```
Вот сюда и нормализацию в рамках одного уравнения добавить

TonitaN commented on course_work_fkm/lib/functions/for_nodes.dart:
```diff
@@ -0,0 +1,119 @@
+  
+  // Ищем самую выгодную подстановку и проводим ее
+  if (substitution.isNotEmpty) {
+    Equation maxBenefit = substitution.entries.reduce((a, b) => a.value > b.value ? a : b).key;
```
По какой метрике самая выгодная?
Метрику, повторюсь, вынести в отдельную функцию

N1X41 committed:
feat: Фиксы

N1X41 committed:
feat: Фикс по тестированию (список в ТГ)
